### PR TITLE
feat(state bridge): gossip contract bytecode and storage

### PR DESCRIPTION
**Pull request by KolbyML**
### What was wrong?

We weren't gossiping contract bytecode and storage, Milos really wanted me to do this before working on being able to execute to the front of the chain. It isn't really that big of a deal, so I am getting this done quick

### How was it fixed?

By adding code to gossip the contract bytecode

For storage because the storage root is updated multiple times per a block, we maintain a cache of updated storage keys per an account.

This implementation is meant to be simple and work, I am aware of optimizations, but those are yet to be seen as needed so we will wait till be run into walls and go from there.

### CODE REVIEW

File: portal-bridge/src/bridge/state.rs
Comment (morph-dev) [inline portal-bridge/src/bridge/state.rs:272]: I think I commented on the other PR, but `AccountProof` should really be called `TrieProof` (or something similar) and it shows here.

**Review by morph-dev — COMMENTED** [review]
I would say no major issues, but will approve once #1318 is merged.

File: portal-bridge/src/bridge/state.rs
Comment (KolbyML) [reply portal-bridge/src/bridge/state.rs:186]: This is a simple/naive implementation that works. We can optimize it when it becomes a bottleneck, but for now gossiping it every block the contract's storage is touched isn't an issue

File: portal-bridge/src/types/state/content.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/content.rs:129]: That doesn't make sense to me as we are getting the address's hash or keccak(addressO

File: portal-bridge/src/types/state/execution.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/execution.rs:46]: I was planning to have a generic config for future use, if you want us to make it more specific now it is fine, I am also fine with keeping it broad and seeing where it goes

File: portal-bridge/src/types/state/config.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/config.rs:5]: We only want this flag enabled if we are gossiping storage changes. I want to be an execution client so by default this feature for our bridge wouldn't be enabled.

**Comment by KolbyML** [comment]
@morph-dev ready for another look let me know if I missed anything

File: portal-bridge/src/types/state/config.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/config.rs:5]: > I'm not sure what you mean by "be an execution client", considering that all this code is inside portal-bridge. Can you explain this a bit more?

This code will be moved into a crate called `trin-execution`. If we are able to execute to the head of the head of the chain we built an execution client. If we are building an execution client (which we are as a byproduct of gossiping state at the head of the chain), then we can release our execution client to the public for others to use. It will be an execution client like Besu, Geth, Reth, Nethermind. The difference is we will be building it without devp2p and will never add devp2p. Devp2p is likely to be dropped sooner or later.

File: portal-bridge/src/bridge/state.rs
Comment (KolbyML) [reply portal-bridge/src/bridge/state.rs:163]: Instead of bail!() I will use continue, as we are not meant to break out the loop, thank you for showing me this cool way to prevent indents

File: portal-bridge/src/bridge/state.rs
Comment (KolbyML) [reply portal-bridge/src/bridge/state.rs:166]: Instead of bail!() I will use continue, as we are not meant to break out the loop, thank you for showing me this cool way to prevent indents

File: portal-bridge/src/types/state/content.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/content.rs:64]: I will do the main suggestion now

and we can do 
> You can also add other useful / common functions to it (e.g. getting the last node, converting path to Nibbles).

In the future if it seems needed

File: portal-bridge/src/types/state/content.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/content.rs:129]: Would this ever be used for something else though? I think we should give it a generic name if we ever use it in a generic way

File: portal-bridge/src/types/state/config.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/config.rs:5]: > Also, I think it would make sense to call it something like gossip_state_changes

This class is only needed for gossip storage trie changes, as the account trie we can gossip it for free. The storage trie's require storing extra stuff in the database, the majority of users won't need this so it is off by default

**Comment by KolbyML** [comment]
@morph-dev ready for another look, let me know if I missed anything

File: portal-bridge/src/bridge/state.rs
Comment (morph-dev) [inline portal-bridge/src/bridge/state.rs:165]: nit: I think we should at least print error here

File: portal-bridge/src/bridge/state.rs
Comment (morph-dev) [inline portal-bridge/src/bridge/state.rs:162]: nit: this comment seems out of place now. Maybe remove it and add appropriate comments where they are needed

File: portal-bridge/src/types/state/content.rs
Comment (morph-dev) [inline portal-bridge/src/types/state/content.rs:34]: I think you should remove this line and simplify this ever further:
```rust
Ok(StateContentValue::AccountTrieNodeWithProof(
    AccountTrieNodeWithProof {
        proof: account_proof.into(),
        block_hash,
    },
))
```

Also in several places below.

File: portal-bridge/src/types/state/content.rs
Comment (morph-dev) [reply portal-bridge/src/types/state/content.rs:104]: I think it's generic enough that I would even extracted into separate file (e.g. `utils.rs`). Reason is that if I'm writing new code and I feel like there might be generic function that I need to use (and I would say this would fall into that category), I'm going to look for it in some library or some `utils.rs` or similar, or by searching by name (and I'm not likely going to find it inside `content.rs` with `full_key_path_to_address_hash` name). Of course this assumes that when you are writing code, you are not familiar with the entire codebase, which is the case in general.

So it is good coding practice to extract generic functions and give them generic names, even if they are currently used only in one place.

I'm not too keen on this, so it's up to you.

Also, you didn't address other comments regarding checking the validity of the input (length can be 63, bytes can have invalid values), which I would might be useful/good even if it is not generic.

File: portal-bridge/src/types/state/config.rs
Comment (morph-dev) [reply portal-bridge/src/types/state/config.rs:5]: > This code will be moved into a crate called `trin-execution`

Yes, this makes sense. It's something I wanted to suggest as well. Thank you for clarifying.

**Review by morph-dev — APPROVED** [review]
:shipit:

File: portal-bridge/src/types/state/content.rs
Comment (KolbyML) [reply portal-bridge/src/types/state/content.rs:104]: > Also, you didn't address other comments regarding checking the validity of the input (length can be 63, bytes can have invalid values), which I would might be useful/good even if it is not generic.

I will add this as I think it is good

Comment (KolbyML) [reply portal-bridge/src/types/state/content.rs:104]: I will move it to utils.rs. I will change the name from `full_key_path_to_address_hash` to ` full_nibble_path_to_address_hash`. I see this function as an inversion to the util function `address_to_nibble_path`. I want to keep `address_hash` in the name because I want when people read `full_nibble_path_to_address_hash` to understand what it is used for. I find `full_nibble_path_to_packed_path` very confusing because the name is so generic I don't understand what it would be used for when reading it.

So if we find a more generic usecase for it in the future we can update it to the generic name. But for now I will do the inbetween name which still contains `address_hash` as you said it is up to me. We still maintain the possibility of changing it depending on what the future brings

File: portal-bridge/src/types/state/utils.rs
Comment (morph-dev) [inline portal-bridge/src/types/state/utils.rs:3]: nit: Most (I would say **all not-trivial**) public functions should have documentation. This also "fixes" the confusion with having more generic names. This greatly helps with coding because of the IDE.

In this case it is more acceptable to keep your name because we have `address_to_nibble_path` next to it.

**Pull request merged**
